### DISCUSSION TITLE
Can't seem to operate on the $response to \React\Http\Browser() 's post()

### DISCUSSION THREAD

**Discussion by mig5** [reactphp/reactphp#518] [Q&A] [answered]
Hi,

I'm running ReactPHP http v1.9.0, async v3.1.0, on PHP 7.4 in an Ubuntu 20.04 VM.

Have some code like this:

```
<?php

use React\EventLoop\Loop;
use function \React\Async\await;

[...]
class MyClass 
{

   public function callback($endpoint, $data)
   {

    try {
        $browser = new \React\Http\Browser();
        foreach ($endpoints as $endpoint) {

            $browser->post(
                $endpoint['url'],
                [
                    'Content-Type' => 'application/json',
                    'Authentication' => xxxxxxxxxxxxxxxxxxxx,
                ],
                json_encode($data)
            )->then(function (Psr\Http\Message\ResponseInterface $response) {
                syslog(LOG_INFO, 'Connection Notification sent to ' . $endpoint[url']);
            }, function (Exception $e) {
                syslog(LOG_DEBUG, 'error on Callback request to ' . $endpoint['url']);
            });
        }
```

It's basically just like https://github.com/clue/reactphp-buzz/blob/master/examples/04-post-json.php

My `post()` request is working great! However, I never am able to surface the `syslog()` call in the `$response` promise. Nothing I put in there ever happens.

Can anyone help me figure out why?

**Comment by SimonFrings** [comment, marked as answer]
Hey @mig5, it seems like you forgot to inherit your `$endpoint` variable inside your closure, should be something like this:

```php
...

)->then(function (Psr\Http\Message\ResponseInterface $response) use ($endpoint) {
    syslog(LOG_INFO, 'Connection Notification sent to ' . $endpoint[url']);
}, function (Exception $e) use ($endpoint) {
    syslog(LOG_DEBUG, 'error on Callback request to ' . $endpoint['url']);
});

....
```

Without the `use ()` PHP won't know what `$endpoint` is, thus PHP will throw and error at this point. This all is happening inside a promise and promises are swallowing exceptions and errors in the current version. We're currently working on changing this behavior to detect all unhandled rejections, more on this in https://github.com/reactphp/promise/issues/87 and https://github.com/orgs/reactphp/discussions/472#discussioncomment-3880271.

Hope this helps :+1:

**Reply by mig5** [reply]
Thanks @SimonFrings ! That's really clear and helpful - and makes a lot of sense!

I can't seem to get it to work, still, but it's probably something I'm doing wrong still...  (I also noted this code (written by someone else) doesn't have an EventLoop attached to the Browser(), and there is no `$loop->run()`.. not sure if significant here (the POST *does* actually work..))

```
            )->then(function (Psr\Http\Message\ResponseInterface $response) use ($endpoint) {
                syslog(LOG_INFO, 'Connection Notification sent!');
            }, function (Exception $e) use ($endpoint) {
                syslog(LOG_DEBUG, 'Error: exception on Callback request: ' . $e->getMessage());
            });
```

I can't even get those syslog events to fire, e.g even if I'm not making use of the `$endpoint` variable in the message.

Is there a quick change I could make to the promise code to surface any potential error, just to check it's even operating on that bit of code?

Thanks again!

**Comment by mig5** [comment]
@SimonFrings just a follow up to say I figured it out. 

My code had other problems in it which weren't related to the imported variable for use in the `then` promise, so the errors were still getting swallowed.

Specifically, in this app, I needed to properly namespace `Psr\Http\Message\ResponseInterface` to be `\Psr\Http\Message\ResponseInterface`, and `Exception` to be `\Exception`.

This code is now working (for both success and exceptions - nice to see HTTP code 4XX, 5XX triggers an Exception here too!):

```
            $browser->post(
                $endpoint['url'],
                [
                    'Content-Type' => 'application/json',
                    'Authentication' => 'xxxxxxxxxxxxxxxxxxxxxxxxx',
                ],
                json_encode($data)
            )->then(function (\Psr\Http\Message\ResponseInterface $response) use ($endpoint) {
                $msg = "Connection Notification sent to " . $endpoint['url'];
                $msg.= ": status " . $response->getStatusCode();
                syslog(LOG_INFO, $msg);
            }, function (\Exception $e) use ($endpoint) {
                $msg = "Error on Callback notification request to " . $endpoint['url'];
                $msg .= ": " . $e->getMessage();
                syslog(LOG_DEBUG, $msg);
            });
```

In the logs:

```
Jul  4 23:41:35 ip-10-101-7-204 php: Error on Callback notification request to https://xxxxxxxxxxxxxxxxxxxx: HTTP status code 504 (Gateway Timeout)

Jul  4 23:41:45 ip-10-101-7-204 php: Connection Notification sent to https://xxxxxxxxxxxxxxxxxxxx status 200

```

Thanks again for your very helpful response! Terrific!

**Reply by SimonFrings** [reply]
@mig5 Thanks for sharing your answer in here, really appreciate you taking the time for this :+1:

Checking the namespaces would've also been my next suggestion, glad it works now  ^^
